docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,310 @@
+represent the conceptual state of a .NET process.
+
+In this document we give a logical description of a data descriptor together with two physical
+manifestations.
+
+The first physical format is used to publish well-known data descriptors in the `dotnet/runtime`
+repository.  It is supposed to be machine- and human-readable.  This format is not meant to be
+particularly consise and may be used for visualization, diagnostics, etc.  Typically data
+descriptors in this form may be written by hand or with the aid of tooling.
+
+The second physical format is used to embed a data descriptor blob within a particularly instance of
+a target runtime.  It is meant to be machine-readable while minimizing the total space needed to
+store it.  It is primarily meant to be read and written by tooling.
+
+## Logical descriptor
+
+Each logical descriptor exists within an implied /target architecture/ consisiting of:
+* target architecture endianness (little endian or big endian)
+* target architecture pointer size (4 bytes or 8 bytes)
+
+The following /primitive types/ are assumed: int8, uint8, int16, uint16, int32, uint32, int64,
+uint64, nint, nuint, pointer, (UTF-8) string.  The multi-byte types are in the target architecture
+endianness.  The types `nint`, `nuint` and `pointer` have target architecture pointer size.
+
+The data descriptor consists of:
+* the data descriptor specification version
+* a collection of type structure descriptors
+* a collection of global value descriptors.
+
+## Data descriptor specification version
+
+This is the version of the phsyical data descriptor.
+
+## Types
+
+The types (both primitive types and structures described by structure descriptors) are classified as
+having either determinate or indeterminate type.  Determinate sizes may be used for pointer
+arithmetic.  Types with indeterminate size may not be.  Note that some sizes may be determinate, but
+/target specific/.  For example pointer types have a fixed size that varies by architecture.
+
+The following primitive types have indeterminate size: `string`.
+
+## Structure descriptors
+
+Each structure descriptor consists of:
+* a name
+* an optional size in bytes
+* a collection of field descriptors
+
+If the size is not given, the type has indeterminate size.  The size may also be given explicitly as
+"indeterminate" to emphasize that the type has indeterminate size.
+
+The collection of field descriptors may be empty.  In that case the type is opaque.  The primitive
+types may be thought of as opaque (for example: on ARM64 `nuint` is an opaque 8 byte type, `int64`
+is another opaque 8 byte type. `string` is an opaque type of indeterminate size).
+
+Type names must be globally unique within a single logical descriptor.
+
+### Field descriptors
+
+Each field descriptor consists of:
+* a name
+* an offset in bytes from the beginning of the struct
+* a type or the special type `variable`
```

Comment:
`MonoType`  is a variable-sized type with extra stuff (encoded custom modifiers) hanging off the end. 

I think we can leave this out for now - I think once we have more experience writing contracts it will be clearer if this is needed or if the algorithmic contracts can meaningfully deal with the trailing variable-sized data.